boskos/janitor/janitor_test.go

```diff
@@ -0,0 +1,193 @@
+func CreateFakeBoskos(resources int) *fakeBoskos {
+	fb := &fakeBoskos{}
+	for i := 0; i < resources; i++ {
+		fb.resources = append(fb.resources,
+			common.Resource{
+				Name:  fmt.Sprintf("res-%d", i),
+				Type:  "project",
+				State: "dirty",
+			})
+	}
+
+	return fb
+}
+
+func (fb *fakeBoskos) Acquire(rtype string, state string, dest string) (string, error) {
+	fb.lock.Lock()
+	defer fb.lock.Unlock()
+
+	for idx := range fb.resources {
+		r := &fb.resources[idx]
+		if r.State == state {
+			r.State = dest
+			fb.wg.Add(1)
+			return r.Name, nil
+		}
+	}
+
+	return "", nil
+}
+
+func (fb *fakeBoskos) ReleaseOne(name string, dest string) error {
+	fb.lock.Lock()
+	defer fb.lock.Unlock()
+
+	for idx := range fb.resources {
+		r := &fb.resources[idx]
+		if r.Name == name {
+			r.State = dest
+			fb.wg.Done()
+			return nil
+		}
+	}
+
+	return fmt.Errorf("No resource %v", name)
+}
+
+// waitTimeout waits for the waitgroup for the specified max timeout.
+// Returns true if waiting timed out.
+func waitTimeout(wg *sync.WaitGroup, timeout time.Duration) bool {
+	c := make(chan struct{})
+	go func() {
+		defer close(c)
+		wg.Wait()
+	}()
+	select {
+	case <-c:
+		return false // completed normally
+	case <-time.After(timeout):
+		return true // timed out
+	}
+}
+
+func TestDrain(t *testing.T) {
+	totalAcquire := 0
+	totalClean := 0
+
+	oldClean := clean
+	defer func() { clean = oldClean }()
+	clean = func(p string) error {
+		totalClean++
+		return nil
+	}
+
+	janitorPool := make(chan string, 1)
+
+	fb := CreateFakeBoskos(5)
+
+	for i := 0; i < 2; i++ {
+		go janitor(fb, janitorPool)
+	}
+
+	for {
+		if proj, err := fb.Acquire("project", "dirty", "cleaning"); err != nil {
+			t.Fatalf("Acquire failed with %v", err)
+		} else if proj == "" {
+			break
+		} else {
+			janitorPool <- proj // will block until pool has a free slot
+			totalAcquire++
+		}
+	}
+
+	if totalAcquire != len(fb.resources) {
+		t.Errorf("Expect to acquire all resources(%d) from fake boskos, got %d", len(fb.resources), totalAcquire)
+	}
+
+	if waitTimeout(&fb.wg, time.Second) {
+		t.Fatal("Expect janitor to finish!")
+	}
+
+	if totalClean != len(fb.resources) {
+		t.Errorf("Expect to clean all resources(%d) from fake boskos, got %d", len(fb.resources), totalAcquire)
+	}
+
+	for _, r := range fb.resources {
+		if r.State != "free" {
+			t.Errorf("Resource %v, expect state free, got state %v", r.Name, r.State)
+		}
+	}
+}
+
+func TestMalfunctionJanitor(t *testing.T) {
+	totalAcquire := 0
+
+	oldClean := clean
+	defer func() { clean = oldClean }()
+	clean = func(p string) error {
+		time.Sleep(time.Hour)
+		return nil
+	}
+
+	janitorPool := make(chan string, 1)
+
+	fb := CreateFakeBoskos(100)
+
+	POOLSIZE = 10
+	for i := 0; i < POOLSIZE; i++ {
+		go janitor(fb, janitorPool)
+	}
+
+	timeout := time.NewTimer(time.Second).C
+
+RUNNER:
+	for {
+		select {
+		case <-timeout:
+			break RUNNER
+		default:
+			if proj, err := fb.Acquire("project", "dirty", "cleaning"); err != nil {
```

Comment:
How about counting the number of things acquired and blow up if this is more than 10?